packages/melody-hooks/src/component.js

```diff
@@ -16,99 +16,273 @@
+    // tracks whether we are in the phase of running effects
+    this.isRunningEffects = false;
+    // stores the data returned from `componentFn`
+    this.data = null;
+
+    // stores the hooks internal configuration and state
     this.hooks = [];
+    // tracks which hook is currently running
     this.hooksPointer = -1;
-    this.isMounted = false;
+
+    // the props the we receive from the parent
     this.props = null;
-    this.state = {};
-    this.data = null;
+    // receiving new props marks the props as dirty
+    this.isPropsDirty = false;
 
+    // the internal state object stores the state
+    // for every `useState` hook by its slot index
+    this.state = Object.create(null);
+    // when `setState` is called we first store
+    // the state in a queue and if we don't see
+    // anymore updates we render the component
+    // with the new state
+    this.stateQueue = Object.create(null);
+    // Calling `setState` will mark the state as dirty.
+    this.isStateDirty = false;
+    // Binds the `setState` function to our component instance
     this.setState = this.setState.bind(this);
+
+    // tracks whether this component is mounted and
+    // attached to the DOM
+    this.isMounted = false;
 }
 
 Object.assign(Component.prototype, {
     /**
      * Set new properties for the Component.
      * This might cause the component to request an update.
+     * @param {*} props new properties
      */
     apply(props) {
         if (shallowEquals(props, this.props)) {
             return;
         }
         this.props = props || null;
+        this.isPropsDirty = true;
         this.enqueueComponent();
     },
 
     /**
      * Executed after a component has been mounted or updated.
-     * After this method has been triggered, the component is considered stable and
-     * accessing the DOM should be safe.
+     * After this method has been triggered, the component is
+     * considered stable and accessing the DOM should be safe.
      * The children of this Component might not have been rendered.
      */
     notify() {
+        if (this.isMounted) {
+            if (options.afterUpdate) afterUpdate(this);
+        } else {
+            this.isMounted = true;
+            if (afterMount) afterMount(this);
+        }
+
         const hooks = this.hooks;
+
+        // Mark this component as in running effects
+        // This is needed in order to not immediately enqueue
+        // the component again when `setState` is called while
+        // we are running the effects
+        this.isRunningEffects = true;
+
+        // Run through hooks and call effects
         for (let i = 0, l = hooks.length; i < l; i++) {
             const hook = hooks[i];
             const type = hook[0];
 
-            switch (type) {
-                case HOOK_TYPE_USE_EFFECT: {
-                    const dirty = hook[3];
-                    if (!dirty) continue;
+            // We are only interested in effect hooks
```

Comment:
Any thoughts on the timing of the effects? As mentioned before, I think we could implement all types of effects in a reasonable way.

Reply:
Yes, I'm not quite sure what's the right solution here.

* `useMutationEffect` immediately after the render function, `setState` calls are forbidden in those effects
* I'm not sure we should have both `useLayoutEffect` and `useEffect`, because if we have `useLayoutEffect` in `notify` and `useEffect` after `notify` inside `requestIdleCallback`, calls to `setState` in `useLayoutEffect` could lead to a new rendering and may start before `useEffect` is finished.

What do you think?

Reply:
In theory we should be able to allow `setState` within both but if it helps we can put in a restriction to say `setState` can not be used within `useMutationEffect` and `useLayoutEffect` for now and then figure out how to do it later.
You'd need some way to ensure scheduled effects will be executed before `render` anyways since there's always a chance of a re-rendering before `requestIdleCallback` or `window.postMessage` trigger.

I do believe it is important to implement `useEffect` with a delay. The exact implementation of the delay is less important at the moment.

Reply:
I still don't understand why it is important to have `useEffect` callbacks invoked with a delay. When `notify` is called, the DOM should be stable.
Doing so would have the following implications: When an update comes in while `useEffect` callbacks are not invoked yet we either:
 * flush those effects to be called before the new update happens (which kinda undermines the idea of having a delay)
 * or we wait for those callbacks to be called and queue the incomming updates (which requires some cumbersome book keeping logic in the component logic)

 Can you explain why it is important?

Reply:
React uses it to ensure a smooth user experience. Basically, the only guarantee they provide you with is "it'll be executed at a time when layout is stable, no rendering is happening and before the next render is executed". That helps them to guarantee a fast render cycle, even if the API user writes horribly slow code.
That's why its useful to have that delay. With the other effect types, the API user can ruin the user experience if they do it wrong. With `useEffect` that's much harder. Worst case you'd slow down scheduling the next render by a bit.
So its easy for the user of the API to write a high performance app without having to deal with `requestIdleCallback` on their own.

Now why would I like to have it in for the first draft instead of adding it later? Because changes to the timing will also change the user experience. Loading some already cached data via a Promise in an Effect might be fast enough if executed right after render (ideally the update happens next frame, so ~16ms later), however, with the delayed/scheduled version the engineer might consider providing a different API for accessing cached data to render it - if present - already during the first render frame instead of waiting more than 50ms (end of current frame + wait for idle time + end of next frame).

On the other hand, reading the options again, I agree neither is ideal. Forcefully flushing before the next render will have a negative impact on that render cycle instead of the previous one and delaying the rendering just to execute some effects also doesn't seem like a solid idea.
I'm wondering why React didn't decide to entirely drop Effects which have been queued, but not executed, in between two renders.

So, to cut things short:
Perhaps we can add the other two effect types (even though they are less relevant for us) and proceed for now. Once we get a better feeling for `useEffect`, we can try to properly integrate it into our `renderQueue`.